### PR TITLE
Add build configs and templatize deployment configs

### PR DESCRIPTION
**Pull request by carbonin**
This PR does a few things.

Primarily it adds build config object templates. This is being done to simplify the process of deploying a new image into the integrated registry within openshift. Previously, we would use a CI hook to push the images, but that meant that the entire registry had to be exposed to the internet which isn't always the case. Additionally developers or QE needed to know how to build and tag all the images and have permissions to push them into the registry. Build configs expose a webhook that we can use to rebuild from github and they also handle tagging and adding the image to the integrated registry.

The build configs in this PR default to adding the newly generated images to the `buildfactory` namespace. This is what will be done in Insights CI and Production so it seemed like a good choice for the default, but if dev or QE wanted to rebuild an image in their own project they can change the namespace easily on both the template and build config now.

Lastly we needed to add image streams as well. Previously, an image stream would be created for us when a tag was pushed to the integrated registry, but with build configs the image stream needs to exist before the build will run.

### CODE REVIEW

**Comment by carbonin** [comment]
There are a few things I've been thinking about combining when creating this PR. In particular, I was thinking we could create a single file for deployment configs/services then also a single file for the build config/image streams. They would still be separate by individual component, but these pairs are so tightly coupled that it might make it easier to just create them at the same time.

Additionally, services and image streams are much less likely to change, so if ops needs to update a deployment config or build config they should be fine overwriting the service or image stream at the same time.

I'm specifically omitting secrets here because we definitely don't want to overwrite those by mistake when patching a deployment.

Thoughts?

**Comment by Fryguy** [comment]
cc @agrare

**Comment by Fryguy** [comment]
> They would still be separate by individual component, but these pairs are so tightly coupled that it might make it easier to just create them at the same time.

This make a lot of sense to me.

File: api/image_stream.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ImageStream
+metadata:
+  name: topological-inventory-api
+spec:
+  tags:
+  - name: latest
+
```
Comment (Fryguy) [inline api/image_stream.yaml]: Extra blank line

File: persister/image_stream.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ImageStream
+metadata:
+  name: topological-inventory-persister
+spec:
+  tags:
+  - name: latest
+
```
Comment (Fryguy) [inline persister/image_stream.yaml]: Same

File: ingress-api/image_stream.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ImageStream
+metadata:
+  name: topological-inventory-ingress-api
+spec:
+  tags:
+  - name: latest
+
```
Comment (Fryguy) [inline ingress-api/image_stream.yaml]: x3 😝

**Comment by bsquizz** [comment]
> In particular, I was thinking we could create a single file for deployment configs/services then also a single file for the build config/image streams.

This is what I've done in e2e-deploy as well ... everything needed to "run the service" is in 1 file. Everything needed to "build the service" (build config, image stream, etc.) is in the other.

Re secrets -- I'd recommend setting your github webhook secret statically -- if your build config is deleted and then re-applied, the generated secret will be different. Ideally with templates we can wipe the env and re-apply it at will. But I believe the value for that parameter will change every single time you run "oc process" on the template file, so you'd be forced to change your config on GitHub...

If you don't want to hard-code the webhook secret into the repo, this is how I handle secrets with ocdeployer & e2e-deploy:

1) The `_cfg.yml` for a service set indicates the service set requires a secret, like:
```
secrets:
- "mysecret1"
- "mysecret2"
```
2) At deploy time, ocdeployer looks in the `--secrets-src-project` to see if it can find a secret with a matching name. If so, it exports the secret from that project and re-imports it into your deployment destination project.
3) Next it will look in the `--secrets-local-dir` to see if it can find a YAML defining a secret with the name. If so, it'll import that secret into your project

So this gives you two options:
* You can store all your secrets in a namespace, which people have limited access to, and at deploy time whoever is running deploy needs to be able to view secrets in that project in order to export them. Perhaps only a Jenkins account used for running deploys and a few humans would actually have access to that project.
* You keep secrets as local files (how you store and obtain them is up to you ...), and import them at deploy time.

So if a secret ever changed, you'd change it in one place, but then you'd re-apply config on all your environments to push the new secret into them.

File: api/build_config_template.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+- apiVersion: v1
+  kind: BuildConfig
+  metadata:
+    name: topological-inventory-api
+  spec:
+    source:
+      type: Git
+      git:
+        uri: git://github.com/ManageIQ/topological_inventory-api
+        ref: master
```
Comment (bsquizz) [inline api/build_config_template.yaml]: Having the ability to specify the URI, ref, and commit with parameters is handy to deploy a custom BuildConfig -- to test a PR or branch for example.

File: api/build_config_template.yaml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    triggers:
+    - type: GitHub
+      github:
+        secret: ${GITHUB_WEBHOOK_SECRET}
+parameters:
+- name: IMAGE_NAMESPACE
+  displayName: Image Namespace
+  description: Namespace to push the resulting image stream tag to
+  value: buildfactory
+- name: GITHUB_WEBHOOK_SECRET
```
Comment (bsquizz) [inline api/build_config_template.yaml]: See my note on secrets

**Comment by carbonin** [comment]
@bsquizz it seems like your secret handling is more geared towards secret objects which would be used for deployments. How does it map to this webhook secret as it's not using a `Secret` object directly?

Also, as for secret value reuse, it feels like we're re-implementing the `--params-file` option to `oc process` wouldn't it be easier to commit a params file to a separate repo then use that in the deploy pipeline?

**Comment by bsquizz** [comment]
@carbonin correct, you'd need to switch to using a Secret object here. That's possible with the github trigger too right?

Good question about `--params-file`, I actually only ever used `--param`. ocdeployer env configs add a little more logic on top of a plain params file. You can set your vars at a global level, per-service-set, or per-component. So I feel like this gives a bit more flexibility than the params file. Then, it also looks at the template you are deploying, and if you define a var which is NOT in the template, it will not pass it to that template when it processes (otherwise oc process yells at you, or at least it does when you use `--param`, haven't tried with `--params-file`). So you can have 1 env file that can be applied to many different service sets. See https://github.com/bsquizz/ocdeployer#environment-file

But yeah -- instead of importing actual 'secret' objects, we could instead just use parameters in the templates (and define the secret objects in the template which use those parameters, if the template needs secret objects), and then just keep the values for those secrets in an env.yml

File: api/build_config_template.yaml (hunk 4 above)
Comment (bsquizz) [reply api/build_config_template.yaml]: Side note, the variables I've been using are:
```
      git:
        ref: ${SOURCE_REPOSITORY_REF}
        uri: ${SOURCE_REPOSITORY_URL}
        commit: ${SOURCE_REPOSITORY_COMMIT}
```

Good question... at the moment only master gets deployed in 'buildfactory' and we only deploy custom builders into other namespaces when we wanna build a specific branch or PR. But yeah I don't think a `${IMAGESTREAM_TAG}` variable, with a default of 'latest' would hurt :) We may want more tags to live in buildfactory one day. For now, just sticking with master/latest.

**Comment by carbonin** [comment]
> you'd need to switch to using a Secret object here. That's possible with the github trigger too right?

Based on the [API docs](https://docs.openshift.com/container-platform/3.9/rest_api/oapi/v1.BuildConfig.html), I'm not sure it is. "secret" in the trigger info is listed as a "string".

**Pull request merged**